torch/optim/sradam.py

```diff
@@ -0,0 +1,91 @@
+import torch
+from .adam import Adam
+
+
+# TODO(crcrpar): Decide whether to override `state_dict` method as
+# this optimizer tracks square root of `exp_avg_sq` and `max_exp_avg_sq`
+# that contradicts with its parent optimizer :class:`torch.optim.Adam`.
+class SRAdam(Adam):
+    r"""Implements Adam algorithm with Stochastic Rounding.
+
+    It has been proposed in `Adam: A Method for Stochastic Optimization`_.
+
+    With Stochastic Rounding, param, `exp_avg`, `exp_avg_sq`, and optionally `max_exp_avg_sq`
+    can be represented with 16 bits.
+    This optimizer requires CUDA.
+
+    Arguments:
+        params (iterable): iterable of parameters to optimize or dicts defining
+            parameter groups
+        lr (float, optional): learning rate (default: 1e-3)
+        betas (Tuple[float, float], optional): coefficients used for computing
+            running averages of gradient and its square (default: (0.9, 0.999))
+        eps (float, optional): term added to the denominator to improve
+            numerical stability (default: 1e-8)
+        weight_decay (float, optional): weight decay (L2 penalty) (default: 0)
+        amsgrad (boolean, optional): whether to use the AMSGrad variant of this
+            algorithm from the paper `On the Convergence of Adam and Beyond`_
+            (default: False)
+
+    .. _Adam\: A Method for Stochastic Optimization:
+        https://arxiv.org/abs/1412.6980
+    .. _On the Convergence of Adam and Beyond:
+        https://openreview.net/forum?id=ryQu7f-RZ
+    """
+
+    _step_supports_amp_scaling = True
+
+    @torch.no_grad()
+    def step(self, closure=None, grad_scaler=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            with torch.no_grad():
+                loss = closure()
+
+        if grad_scaler is not None:
```

Comment:
is `torch.cuda.current_device()` OK here? Could this create a CUDA context and allocate memory on GPU0, if we manually used GPU1 for all ops?

Reply:
So is it safer to get `get_inf` per parameter (or group) by `found_inf = device_found_inf_dict[param.device]`?

Reply:
Would it be possible to use the device of `scale` or another param?

Reply:
`GradScaler._check_inf_per_device` populates a different `found_inf` tensor on every device where the optimizer's parameters are.  To make the the decision about whether to skip the step, `GradScaler.step` [combines](https://github.com/pytorch/pytorch/blob/master/torch/cuda/amp/grad_scaler.py#L283) the `found_inf` tensors from all devices.

`_check_inf_per_device` does NOT automatically combine the `found_inf` tensors from all devices.  Your current code passes each device's `found_inf` tensor to the `stochastic_rounding_adam_step` kernel separately, which is incorrect.  If infs are found for params on one device but not others, the stochastic rounding kernel only skips the math for params on the device where infs were found.

Your logic should combine the `found_inf` tensors from all devices, then wrap the combined found inf tensor in `_MultiDeviceReplicator` for later use by the kernel.

A similar fix will be required in the other `SR*` optimizers.  In fact this requirement+logic are generic across optimizers.  You could make a helper function to wrap the logic above (that returns a replicator that wraps a combined found inf tensor), then all optimizers could call the helper function.  Here's an example that remains sync-free in the common case (logic is similar to [GradScaler.update](https://github.com/pytorch/pytorch/blob/1039b95ff056ba0e0118d40267c68a9df4535f7b/torch/cuda/amp/grad_scaler.py#L325-L334)):
```python
def _combined_found_inf_helper(optimizer, scaler):
    found_inf_dict = grad_scaler._check_inf_per_device(optimizer)
    # Combines found_inf tensors from all devices.  As in GradScaler.update(),
    # tensors are combined on the scale's device, which is an arbitrary but 
    # reasonable choice that avoids new context creation.
    found_infs = [f.to(scaler._scale.device, non_blocking=True) for f in found_inf_dict.values()]
    assert len(found_infs) > 0, "No inf checks were recorded in _check_inf_per_device."
    found_inf_combined = found_infs[0]
    if len(found_infs) > 1:
        for i in range(1, len(found_infs)):
            found_inf_combined += found_infs[i]
    return _MultiDeviceReplicator(found_inf_combined)
```
then within an optimizer e.g.
```
found_inf = _combined_found_inf_helper(self)
...
# same kernel call, but now it's safe, because found_inf accounts for infs from all devices
torch.stochastic_rounding_adam_step(..., found_inf.get(param.device), ...) 
```

Reply:
Thank you for the detailed and easy to understand explanation and suggestion!!!